FOGL-3516: Add sub-second granularity to time bucket size

pintomax opened the pull request:
FOGL-3516: use julianday instaead of strftime

pintomax committed:
FOGL-3516: use julianday instaead of strftime

FOGL-3516: use julianday instaead of strftime

pintomax committed:
Merge branch 'develop' into FOGL-3516

pintomax committed:
FOGL-3516: Initial implementation of sub-second granularity to time bucket size

FOGL-3516: Initial implementation of sub-second granularity to time bucket size

pintomax committed:
Fixed comma and space in SQL

Fixed comma and space in SQL

pintomax commented:
Examples:
50 ms
curl -s 'localhost:8081/foglamp/asset/sinusoid/bucket/0.05?length=16000&start=1576244444'| jq
``` 
 {
    "asset_code": "sinusoid",
    "timestamp": "2019-12-13 14:58:24.372467",
    "reading": {
      "sinusoid": {
        "min": 0.743144825,
        "max": 0.951056516,
        "average": 0.8565578394,
        "count": 5,
        "sum": 4.282789197
      }
    }
  },
  {
    "asset_code": "sinusoid",
    "timestamp": "2019-12-13 14:58:24.322500",
    "reading": {
      "sinusoid": {
        "min": 0.309016994,
        "max": 0.669130606,
        "average": 0.494533899,
        "count": 5,
        "sum": 2.472669495
      }
    }
  },
  {
    "asset_code": "sinusoid",
    "timestamp": "2019-12-13 14:58:24.272467",
    "reading": {
      "sinusoid": {
        "min": 0,
        "max": 0.207911691,
        "average": 0.104146718,
        "count": 3,
        "sum": 0.312440154
      }
    }
  }
```

500 ms

curl -s localhost:8081/foglamp/asset/sinusoid/bucket/0.5?length=16000| jq
```
  {
    "asset_code": "sinusoid",
    "timestamp": "2019-12-13 14:58:38.745906",
    "reading": {
      "sinusoid": {
        "min": -1,
        "max": 0.669130606,
        "average": -0.382306582609756,
        "count": 41,
        "sum": -15.674569887
      }
    }
  },
  {
    "asset_code": "sinusoid",
    "timestamp": "2019-12-13 14:58:38.249764",
    "reading": {
      "sinusoid": {
        "min": -1,
        "max": 1,
        "average": 0.200455028428571,
        "count": 42,
        "sum": 8.419111194
      }
    }
  }
```

pintomax committed:
Allow float fo be passed to asset_datapoints_with_bucket_size

Allow float fo be passed to asset_datapoints_with_bucket_size

pintomax commented:
NOTE: Float value is only passed to asset_datapoints_with_bucket_size()

pintomax committed:
Merge branch 'develop' into FOGL-3516

pintomax committed:
Handle subsecond in start parameter

Handle subsecond in start parameter

pintomax committed:
Legth parameter can be a float number

Legth parameter can be a float number

pintomax committed:
Microseconds in date are in use for both length ans start parameters

Microseconds in date are in use for both length ans start parameters

pintomax committed:
Build UTC datetime for storage layer queries

Build UTC datetime for storage layer queries

MarkRiddoch commented on C/plugins/storage/sqlite/common/readings.cpp:
```diff
@@ -186,28 +200,60 @@ bool Connection::aggregateQuery(const Value& payload, string& resultSet)
-		sql.append(") / ");
-		sql.append(to_string(size));
-		sql.append(", 6), 'unixepoch') ");
+		sql.append(") - 2440587.5) * 86400.0 / ");
```
It might be good to use #define's or constants here, just so everybody is clear what 2440587.5 and 86400.0 re - my guess is they are some important numbers!

MarkRiddoch commented on C/plugins/storage/sqlite/common/readings.cpp:
```diff
@@ -186,28 +200,60 @@ bool Connection::aggregateQuery(const Value& payload, string& resultSet)
+		sql.append("round((julianday(");
 		sql.append(timeColumn);
-		sql.append(") / 1, 6), 'unixepoch') ");
+		sql.append(") - 2440587.5) * 86400.0 / 1)");
```
Constants again

MarkRiddoch commented on C/plugins/storage/sqlite/common/readings.cpp:
```diff
@@ -220,28 +266,27 @@ bool Connection::aggregateQuery(const Value& payload, string& resultSet)
 	sql.append(timeColumn);
-	sql.append(") / ");
-	if (size > 1)
+	sql.append(") - 2440587.5) * 86400.0 / ");
```
These numbers are clearly important!

MarkRiddoch reviewed (changes requested):
Just a vey minor issue with the use of constants

pintomax committed:
Fix for start parameter

Fix for start parameter

pintomax committed:
Merge branch 'develop' into FOGL-3516

pintomax committed:
Added constants needed fro JulianDay to UnixTimestamp conversion

Added constants needed fro JulianDay to UnixTimestamp conversion

pintomax committed:
Fix for float string representation

Fix for float string representation

It was
"0.00005" ==> "5e-05"

Now it is ==>" 0.000050"

shariq1501 commented:
```
http://localhost:8081/foglamp/asset/sinusoid/bucket/1?length=600
Response:
400 'NoneType' object is not subscriptable
```

pintomax committed:
Fix SQL statement build

Fix SQL statement build

pintomax committed:
Fix microseconds addition in built dates

Fix microseconds addition in built dates

pintomax committed:
Updated SQLite3 unit tests

Updated SQLite3 unit tests

Usage of julianday (with microseconds) affects how display time in bucket summary is reported.

Display time (2019-10-11 15:11:00) is now close to max time (2019-10-11 15:10:54) in dataset.

Previous value was "2019-10-11 15:10:40"

pintomax committed:
Update Postgres storage pòugin for subsecond bucket

Update Postgres storage pòugin for subsecond bucket

pintomax committed:
Postgres unit tests updated: bucket query uses round instead of floor

Postgres unit tests updated: bucket query uses round instead of floor

MarkRiddoch reviewed (dismissed):
Looks good as far as I can tell

pintomax committed:
Fix Buster compilation issue

Fix Buster compilation issue
